components/brave_search_conversion/p3a.cc

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_search_conversion/p3a.h"
+
+#include "base/logging.h"
+#include "base/metrics/histogram_functions.h"
+#include "brave/components/brave_search_conversion/pref_names.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+
+namespace brave_search_conversion {
+namespace p3a {
+
+namespace {
+
+const char* GetOmniboxShownPrefName(ConversionType type) {
+  switch (type) {
+    case ConversionType::kBanner:
+      return prefs::kP3ABannerShown;
+    case ConversionType::kButton:
+      return prefs::kP3AButtonShown;
+    default:
+      return nullptr;
+  }
+}
+
+const char* GetOmniboxTriggeredPrefName(ConversionType type) {
+  switch (type) {
+    case ConversionType::kBanner:
+      return prefs::kP3ABannerTriggered;
+    case ConversionType::kButton:
+      return prefs::kP3AButtonTriggered;
+    default:
+      return nullptr;
+  }
+}
+
+const char* GetPromoTypeHistogramName(ConversionType type) {
+  switch (type) {
+    case ConversionType::kBanner:
+      return kSearchPromoBannerHistogramName;
+    case ConversionType::kButton:
+      return kSearchPromoButtonHistogramName;
+    default:
+      return nullptr;
+  }
+}
+
+void UpdateHistograms(PrefService* prefs) {
+  const ConversionType types[] = {ConversionType::kBanner,
+                                  ConversionType::kButton};
+
+  VLOG(1) << "SearchConversionP3A: updating histograms";
+
+  const bool default_engine_triggered =
+      prefs->GetBoolean(prefs::kP3ADefaultEngineChanged);
+  for (const auto type : types) {
+    const char* shown_pref_name = GetOmniboxShownPrefName(type);
+    const char* triggered_pref_name = GetOmniboxTriggeredPrefName(type);
+    const char* histogram_name = GetPromoTypeHistogramName(type);
+    DCHECK(shown_pref_name);
+    DCHECK(triggered_pref_name);
+    DCHECK(histogram_name);
+    if (!prefs->GetBoolean(shown_pref_name)) {
```

Comment:
I believe this will still crash in release if `Get...Name()` returns `nullptr`. There's no danger as written, but it would be safer to return an empty string `""`, and `DCHECK(..._name.length())` instead. That way constructing the `std::string` temporary would still succeed but `PrefService::GetBoolean()` would return false.